File: packages/ingest/src/rpg_ingest/raw/reading_order.py
from __future__ import annotations

import re
import unicodedata

from rpg_ingest.raw.layout import LayoutBlock, LayoutPage

SPATIAL_Y_TOLERANCE = 5.0
MIN_COLUMN_OVERLAP = 0.35
NARROW_BOX_MAX_WIDTH = 160.0
NARROW_BOX_X_MARGIN = 35.0
NARROW_BOX_MAX_VERTICAL_GAP = 130.0
DECORATIVE_FONT_RATIO = 2.0
DECORATIVE_MIN_FONT = 28.0
DECORATIVE_TOP_RATIO = 0.33
VERTICAL_HEADER_MAX_WIDTH = 20.0
VERTICAL_HEADER_MIN_X_RATIO = 0.85
TITLE_CASE_MAX_WORDS = 6
TITLE_CASE_MIN_WORDS = 2

ALL_CAPS_RE = re.compile(r"^[A-Z0-9ÀÂÄÉÈÊËÏÎÔÙÛÜŸÇ][A-Z0-9ÀÂÄÉÈÊËÏÎÔÙÛÜŸÇ\s\-:,'\.]{2,}$")
TITLE_CASE_WORD_RE = re.compile(
    r"^[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜŸÇ][\w'\u2019\-]+"
    r"(?:\s+(?:[a-zàâäéèêëïîôùûüÿç][\w'\u2019\-]+|[A-Z]{2,})){1,5}$",
    re.UNICODE,
)
PAGE_NUMBER_LABEL_RE = re.compile(r"^PAGE\s+\d+\s*$", re.IGNORECASE)
CHAPTER_RE = re.compile(
    r"^(?:chapter|chapitre|part|partie)\s+(\d+|[IVXLC]+)\b",
    re.IGNORECASE,
)
LIST_ITEM_MARKER_RE = re.compile(
    r"^[\u0090-\u0095\u2022\u25AA-\u25C6\u25E6\-–—]\s*",
)
LIST_ITEM_NAME_RE = re.compile(
    r"^[\wÀ-ÿ][\wÀ-ÿ\s'\-]{0,40}\u202f:",
)

META_BOX_HEADINGS = frozenset(
    {
        "CRÉDITS",
        "EN QUELQUES MOTS",
        "FICHE TECHNIQUE",
    }
)

EDITORIAL_CREDITS_MARKERS = (
    "black book",
    "tous droits réservés",
    "tous droits reserves",
)

MAX_SUBORDINATE_CHAPTER_PAGE_GAP = 3

PAGE_BANNER_PREFIXES = ("INTRODUCTION", "IMPLICATION", "CONCLUSION")

CONDITIONAL_HOOK_RE = re.compile(r"^Si\s+", re.IGNORECASE)
CONDITIONAL_HOOK_MAX_WORDS = 10


def _strip_glyphs(text: str) -> str:
    return "".join(
        ch for ch in text if unicodedata.category(ch) not in {"Cf", "Co", "Cs"}
    ).strip()


def spatial_sort_key(block: LayoutBlock) -> tuple[int, float, float]:
    y_bucket = round(block.bbox.y0 / SPATIAL_Y_TOLERANCE) * SPATIAL_Y_TOLERANCE
    return (block.page_number, y_bucket, block.bbox.x0)


def is_same_y_band(left: LayoutBlock, right: LayoutBlock) -> bool:
    left_bucket = round(left.bbox.y0 / SPATIAL_Y_TOLERANCE) * SPATIAL_Y_TOLERANCE
    right_bucket = round(right.bbox.y0 / SPATIAL_Y_TOLERANCE) * SPATIAL_Y_TOLERANCE
    return left_bucket == right_bucket


def column_side(block: LayoutBlock, page_width: float) -> str:
    center = (block.bbox.x0 + block.bbox.x1) / 2
    return "left" if center < page_width / 2 else "right"


def column_major_sort_key(page: LayoutPage, block: LayoutBlock) -> tuple[int, int, float, float]:
    side = 0 if column_side(block, page.width) == "left" else 1
    return (block.page_number, side, block.bbox.y0, block.bbox.x0)


def horizontal_overlap_ratio(left: LayoutBlock, right: LayoutBlock) -> float:
    overlap = min(left.bbox.x1, right.bbox.x1) - max(left.bbox.x0, right.bbox.x0)
    if overlap <= 0:
        return 0.0
    narrower = min(left.bbox.x1 - left.bbox.x0, right.bbox.x1 - right.bbox.x0)
    if narrower <= 0:
        return 0.0
    return overlap / narrower


def is_in_column_band(
    block: LayoutBlock,
    heading: LayoutBlock,
    *,
    min_overlap: float = MIN_COLUMN_OVERLAP,
) -> bool:
    return horizontal_overlap_ratio(block, heading) >= min_overlap


def is_narrow_heading_box(heading: LayoutBlock) -> bool:
    return (heading.bbox.x1 - heading.bbox.x0) <= NARROW_BOX_MAX_WIDTH


def is_in_heading_content_zone(
    block: LayoutBlock,
    heading: LayoutBlock,
    *,
    heading_text: str | None = None,
    min_overlap: float = MIN_COLUMN_OVERLAP,
) -> bool:
    # Reject only blocks entirely above the heading. In bi-column layouts the
    # parallel column often starts slightly higher than the heading box.
    if block.bbox.y1 <= heading.bbox.y0:
        return False
    if horizontal_overlap_ratio(block, heading) < min_overlap:
        return False
    if heading_text and is_meta_box_heading(heading_text):
        if block.bbox.x0 < heading.bbox.x0 - NARROW_BOX_X_MARGIN:
            return False
        if block.bbox.x1 > heading.bbox.x1 + NARROW_BOX_X_MARGIN:
            return False
        gap = block.bbox.y0 - heading.bbox.y1
        return 0 <= gap <= NARROW_BOX_MAX_VERTICAL_GAP
    return True


def page_median_font(blocks: list[LayoutBlock]) -> float:
    sizes = [
        b.metadata.get("max_font_size")
        for b in blocks
        if b.metadata.get("max_font_size")
    ]
    if not sizes:
        return 12.0
    sizes.sort()
    mid = len(sizes) // 2
    return sizes[mid] if len(sizes) % 2 else (sizes[mid - 1] + sizes[mid]) / 2


def is_decorative_spread_title(
    block: LayoutBlock,
    page: LayoutPage,
    *,
    median_font: float,
) -> bool:
    text = _strip_glyphs(block.text).strip()
    if not text or not ALL_CAPS_RE.match(text):
        return False
    max_font = block.metadata.get("max_font_size") or 0
    if (
        max_font < DECORATIVE_MIN_FONT
        and max_font < median_font * DECORATIVE_FONT_RATIO
    ):
        return False
    if block.bbox.y0 > page.height * DECORATIVE_TOP_RATIO:
        return False
    return len(text.split()) <= 4 and text.count("\n") <= 1


def is_spread_title_pair(
    upper: LayoutBlock,
    lower: LayoutBlock,
    page: LayoutPage,
    *,
    median_font: float,
) -> bool:
    if not is_decorative_spread_title(upper, page, median_font=median_font):
        return False
    lower_text = _strip_glyphs(lower.text).strip()
    if not lower_text or not ALL_CAPS_RE.match(lower_text):
        return False
    if lower.bbox.y0 > page.height * DECORATIVE_TOP_RATIO:
        return False
    if lower.bbox.y0 < upper.bbox.y1 - 5:
        return False
    return len(lower_text.split()) <= 4


def is_vertical_running_header(block: LayoutBlock, page: LayoutPage) -> bool:
    width = block.bbox.x1 - block.bbox.x0
    if width > VERTICAL_HEADER_MAX_WIDTH:
        return False
    if block.bbox.x0 < page.width * VERTICAL_HEADER_MIN_X_RATIO:
        return False
    text = _strip_glyphs(block.text).strip()
    return bool(text) and len(text) <= 80 and len(text.split()) <= 12


def is_page_number_label(text: str) -> bool:
    return bool(PAGE_NUMBER_LABEL_RE.match(_strip_glyphs(text).strip()))


def is_page_number_label_block(block: LayoutBlock) -> bool:
    return is_page_number_label(block.text)


def is_page_footer_block(block: LayoutBlock, page: LayoutPage, *, footer_ratio: float = 0.08) -> bool:
    footer_limit = page.height * (1.0 - footer_ratio)
    if block.bbox.y0 <= footer_limit:
        return False
    return is_page_number_label(block.text)


def is_page_banner_heading(
    text: str,
    block: LayoutBlock,
    page: LayoutPage,
) -> bool:
    cleaned = _strip_glyphs(text).strip()
    if not cleaned:
        return False
    first_line = cleaned.split("\n")[0].strip()
    upper = first_line.upper()
    if not any(upper.startswith(prefix) for prefix in PAGE_BANNER_PREFIXES):
        return False
    if block.bbox.y0 > page.height * DECORATIVE_TOP_RATIO:
        return False
    letters = [char for char in first_line if char.isalpha()]
    if not letters:
        return False
    upper_ratio = sum(1 for char in letters if char.isupper()) / len(letters)
    return upper_ratio >= 0.8


def is_conditional_hook_heading(
    text: str,
    block: LayoutBlock,
    *,
    median_font: float,
) -> bool:
    cleaned = _strip_glyphs(text).strip()
    if not cleaned or not CONDITIONAL_HOOK_RE.match(cleaned):
        return False
    if len(cleaned.split()) > CONDITIONAL_HOOK_MAX_WORDS:
        return False
    if not block.metadata.get("is_bold"):
        return False
    max_font = block.metadata.get("max_font_size") or 0
    if max_font >= median_font * 1.5:
        return False
    return max_font >= median_font * 1.05


def is_title_case_heading(
    text: str,
    block: LayoutBlock,
    *,
    median_font: float,
) -> bool:
    cleaned = _strip_glyphs(text).strip()
    if not cleaned or not block.metadata.get("is_bold"):
        return False
    max_font = block.metadata.get("max_font_size") or 0
    if max_font < median_font * 1.05:
        return False
    if is_conditional_hook_heading(text, block, median_font=median_font):
        return True
    words = cleaned.split()
    if len(words) == 1:
        return (
            block.metadata.get("is_bold", False)
            and max_font >= median_font * 1.05
            and cleaned[0].isupper()
            and any(char.islower() for char in cleaned[1:])
        )
    if not (TITLE_CASE_MIN_WORDS <= len(words) <= TITLE_CASE_MAX_WORDS):
        return False
    return bool(TITLE_CASE_WORD_RE.match(cleaned))


def is_meta_box_heading(text: str) -> bool:
    normalized = _strip_glyphs(text).strip().upper()
    return normalized in META_BOX_HEADINGS


def is_credits_heading(text: str) -> bool:
    return _strip_glyphs(text).strip().upper() == "CRÉDITS"


def is_editorial_credits_text(text: str) -> bool:
    lowered = _strip_glyphs(text).lower()
    return any(marker in lowered for marker in EDITORIAL_CREDITS_MARKERS)


def is_editorial_credits_block(block: LayoutBlock) -> bool:
    return is_editorial_credits_text(block.text)


def normalize_section_title(text: str) -> str:
    stripped = _strip_glyphs(text).strip()
    if "\n" in stripped and is_all_caps_heading_text(stripped):
        return " ".join(stripped.split())
    return stripped


def is_encadre_title_line(block: LayoutBlock) -> bool:
    text = _strip_glyphs(block.text).strip()
    if not text or not block.metadata.get("is_bold"):
        return False
    first_line = text.split("\n", 1)[0].strip()
    if len(first_line) < 4 or not first_line.isupper():
        return False
    return len(first_line.split()) <= 12


def is_all_caps_heading_text(text: str) -> bool:
    first_line = _strip_glyphs(text).split("\n")[0].strip()
    return bool(ALL_CAPS_RE.match(first_line))


def is_chapter_heading(text: str) -> bool:
    return bool(CHAPTER_RE.match(_strip_glyphs(text).strip()))


def is_list_item_block(block: LayoutBlock) -> bool:
    text = _strip_glyphs(block.text).strip()
    if not text:
        return False
    if LIST_ITEM_MARKER_RE.match(text):
        return True
    if block.metadata.get("is_bold") and len(text.split()) <= 6:
        return False
    if LIST_ITEM_NAME_RE.match(text) and len(text) <= 80:
        return True
    return False


def heading_visual_tier(
    text: str,
    block: LayoutBlock,
    *,
    median_font: float,
    page: LayoutPage | None = None,
) -> str:
    """Classify heading style: meta box, chapter/part, page banner, subordinate, or other."""
    if is_meta_box_heading(text):
        return "meta"
    if page is not None and is_page_banner_heading(text, block, page):
        return "banner"
    if is_chapter_heading(text):
        return "chapter"
    if is_title_case_heading(text, block, median_font=median_font):
        return "subordinate"
    return "other"


def is_subordinate_to_chapter(
    text: str,
    block: LayoutBlock,
    *,
    median_font: float,
) -> bool:
    return heading_visual_tier(text, block, median_font=median_font) == "subordinate"


def find_block(
    pages: list[LayoutPage], page_number: int, block_index: int
) -> LayoutBlock | None:
    for page in pages:
        if page.page_number != page_number:
            continue
        for block in page.blocks:
            if block.block_index == block_index:
                return block
    return None


def spatially_sorted_headings(
    headings: list[tuple[int, int, str, int]],
    pages: list[LayoutPage],
) -> list[tuple[int, int, str, int]]:
    def sort_key(item: tuple[int, int, str, int]) -> tuple[int, float, float]:
        page_num, block_idx, _, _ = item
        block = find_block(pages, page_num, block_idx)
        if block is None:
            return (page_num, float(block_idx), 0.0)
        return spatial_sort_key(block)

    return sorted(headings, key=sort_key)


def page_is_sparse(page: LayoutPage, *, max_blocks: int = 2) -> bool:
    body_blocks = [
        b
        for b in page.blocks
        if not is_page_number_label_block(b)
        and not is_vertical_running_header(b, page)
    ]
    return len(body_blocks) <= max_blocks

File: packages/ingest-clj/src/rpg/ingest/reading_order.clj
(ns rpg.ingest.reading-order
  "Ordre de lecture (passe 1) et utilitaires typo/géométrie (passe 2)."
  (:require [clojure.string :as str]))

(def ^:private min-column-overlap 0.35)
(def ^:private narrow-box-max-width 160.0)
(def ^:private narrow-box-x-margin 35.0)
(def ^:private narrow-box-max-vertical-gap 130.0)
(def ^:private decorative-font-ratio 2.0)
(def ^:private decorative-min-font 28.0)
(def ^:private decorative-top-ratio 0.33)
(def ^:private vertical-header-max-width 20.0)
(def ^:private vertical-header-min-x-ratio 0.85)
(def ^:private title-case-max-words 6)
(def ^:private title-case-min-words 2)
(def ^:private conditional-hook-max-words 10)
(def ^:private caps-subordinate-max-gap 80.0)

(def max-subordinate-chapter-page-gap 3)

(def ^:private page-banner-prefixes
  ["INTRODUCTION" "IMPLICATION" "CONCLUSION"])

(def ^:private meta-box-headings
  #{"CRÉDITS" "EN QUELQUES MOTS" "FICHE TECHNIQUE"})

(def ^:private editorial-credits-markers
  ["black book" "tous droits réservés" "tous droits reserves"])

(def ^:private all-caps-re
  #"^[A-Z0-9ÀÂÄÉÈÊËÏÎÔÙÛÜŸÇ][A-Z0-9ÀÂÄÉÈÊËÏÎÔÙÛÜŸÇ\s\-:,'\.]{2,}$")

(def ^:private title-case-word-re
  #"(?u)^[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜŸÇ][\w'\u2019\-]+(?:\s+(?:[a-zàâäéèêëïîôùûüÿç][\w'\u2019\-]+|[A-Z]{2,})){1,5}$")

(def ^:private page-number-label-re
  #"(?i)^PAGE\s+\d+\s*$")

(def ^:private chapter-re
  #"(?i)^(?:chapter|chapitre|part|partie)\s+(\d+|[IVXLC]+)\b")

(def ^:private list-item-marker-re
  #"^[\u0090-\u0095\u2022\u25AA-\u25C6\u25E6\-–—]\s*")

(def ^:private list-item-name-re
  #"(?u)^[\wÀ-ÿ][\wÀ-ÿ\s'\-]{0,40}\u202f:")

(def ^:private numbered-heading-re
  #"^(\d+(?:\.\d+)*)\s+(.+)$")

(def ^:private table-row-label-re
  #"^[A-Z]-\d+$")

(def ^:private conditional-hook-re
  #"(?i)^Si\s+")

(defn spatial-sort-key
  "Tri générique : position horizontale puis verticale."
  [block]
  [(get-in block [:bbox :x0])
   (get-in block [:bbox :y0])
   (get-in block [:bbox :x1])])

(defn sort-blocks-spatial
  [blocks]
  (vec (sort-by spatial-sort-key blocks)))

(defn reindex-blocks
  "Réattribue :block-index 0..n dans l'ordre de la liste."
  [blocks]
  (vec (map-indexed (fn [idx block] (assoc block :block-index idx)) blocks)))

(defn normalize-page-blocks
  "Trie les blocs en ordre spatial et ré-indexe."
  [blocks]
  (-> blocks sort-blocks-spatial reindex-blocks))

(defn normalize-page
  [page]
  (update page :blocks normalize-page-blocks))

(defn normalize-reading-order
  "Applique le tri spatial sur chaque page."
  [pages]
  (mapv normalize-page pages))

(defn spatial-ordered?
  "Vérifie qu'une liste de blocs respecte l'ordre (x0, y0, x1)."
  [blocks]
  (every? true?
          (map (fn [[prev cur]]
                 (neg? (compare (spatial-sort-key prev) (spatial-sort-key cur))))
               (partition 2 1 blocks))))

(defn horizontal-overlap-ratio
  "Ratio de chevauchement horizontal (plus étroit / overlap) — pour passe 2."
  [left right]
  (let [overlap (- (min (get-in left [:bbox :x1]) (get-in right [:bbox :x1]))
                   (max (get-in left [:bbox :x0]) (get-in right [:bbox :x0])))]
    (if (<= overlap 0)
      0.0
      (let [narrower (min (- (get-in left [:bbox :x1]) (get-in left [:bbox :x0]))
                          (- (get-in right [:bbox :x1]) (get-in right [:bbox :x0])))]
        (if (<= narrower 0)
          0.0
          (/ overlap narrower))))))

(defn is-in-column-band?
  "Deux blocs partagent-ils la même bande horizontale (fils de lecture parallèles)."
  ([block heading]
   (is-in-column-band? block heading min-column-overlap))
  ([block heading min-overlap]
   (>= (horizontal-overlap-ratio block heading) min-overlap)))

(defn same-reading-stream?
  "Même fil de lecture que le bloc de référence."
  [ref-block block]
  (is-in-column-band? block ref-block))

(defn block-x-cluster
  "Cluster x grossier (x0) pour assertions de test."
  [block]
  (get-in block [:bbox :x0]))

(defn- layout-char? [c]
  (let [ch (char (int c))
        t (Character/getType ch)]
    (or (= t Character/FORMAT)
        (= t Character/PRIVATE_USE)
        (= t Character/SURROGATE))))

(defn strip-glyphs
  [text]
  (->> text
       (remove layout-char?)
       (apply str)
       str/trim))

(defn block-max-font
  [block]
  (double (or (get-in block [:metadata :max-font-size])
              (get-in block [:metadata :max_font_size])
              0.0)))

(defn block-bold?
  [block]
  (boolean (or (get-in block [:metadata :is-bold])
               (get-in block [:metadata :is_bold]))))

(defn column-side
  [block page-width]
  (let [center (/ (+ (get-in block [:bbox :x0]) (get-in block [:bbox :x1])) 2.0)]
    (if (< center (/ page-width 2.0)) "left" "right")))

(defn column-major-sort-key
  [page block]
  (let [side (if (= "left" (column-side block (:width page))) 0 1)]
    [(:page-number page) side (get-in block [:bbox :y0]) (get-in block [:bbox :x0])]))

(defn page-median-font
  [blocks]
  (let [sizes (->> blocks
                   (map block-max-font)
                   (filter pos?)
                   sort
                   vec)]
    (if (empty? sizes)
      12.0
      (let [mid (quot (count sizes) 2)]
        (if (odd? (count sizes))
          (nth sizes mid)
          (/ (+ (nth sizes (dec mid)) (nth sizes mid)) 2.0))))))

(defn is-decorative-spread-title?
  [block page median-font]
  (let [text (strip-glyphs (:text block))
        max-font (block-max-font block)]
    (and (seq text)
         (re-matches all-caps-re text)
         (or (>= max-font decorative-min-font)
             (>= max-font (* median-font decorative-font-ratio)))
         (<= (get-in block [:bbox :y0]) (* (:height page) decorative-top-ratio))
         (<= (count (str/split text #"\s+")) 4)
         (<= (count (str/split text #"\n")) 2))))

(defn- spread-title-vertical-gap [upper lower]
  (- (get-in lower [:bbox :y0]) (get-in upper [:bbox :y1])))

(defn is-spread-title-pair?
  [upper lower page median-font]
  (let [lower-text (strip-glyphs (:text lower))
        gap (spread-title-vertical-gap upper lower)]
    (and (is-decorative-spread-title? upper page median-font)
         (seq lower-text)
         (re-matches all-caps-re lower-text)
         (<= (get-in lower [:bbox :y0]) (* (:height page) decorative-top-ratio))
         (<= gap 8.0)
         (>= gap -30.0)
         (<= (count (str/split lower-text #"\s+")) 4))))

(defn is-vertical-running-header?
  [block page]
  (let [width (- (get-in block [:bbox :x1]) (get-in block [:bbox :x0]))
        text (strip-glyphs (:text block))]
    (and (<= width vertical-header-max-width)
         (>= (get-in block [:bbox :x0]) (* (:width page) vertical-header-min-x-ratio))
         (seq text)
         (<= (count text) 80)
         (<= (count (str/split text #"\s+")) 12))))

(defn is-page-number-label?
  [text]
  (boolean (re-matches page-number-label-re (strip-glyphs text))))

(defn is-page-banner-heading?
  [text block page]
  (let [cleaned (strip-glyphs text)
        first-line (-> cleaned (str/split #"\n") first str/trim)
        upper (str/upper-case first-line)
        letters (filter #(Character/isLetter ^char %) first-line)]
    (and (seq cleaned)
         (some #(str/starts-with? upper %) page-banner-prefixes)
         (<= (get-in block [:bbox :y0]) (* (:height page) decorative-top-ratio))
         (seq letters)
         (>= (/ (count (filter #(Character/isUpperCase ^char %) letters))
               (count letters))
             0.8))))

(defn is-conditional-hook-heading?
  [text block median-font]
  (let [cleaned (strip-glyphs text)
        max-font (block-max-font block)]
    (and (seq cleaned)
         (re-matches conditional-hook-re cleaned)
         (<= (count (str/split cleaned #"\s+")) conditional-hook-max-words)
         (block-bold? block)
         (< max-font (* median-font 1.5))
         (>= max-font (* median-font 1.05)))))

(defn is-title-case-heading?
  [text block median-font]
  (let [cleaned (-> text strip-glyphs (str/split #"\s+") (->> (str/join " ")))
        max-font (block-max-font block)
        words (str/split cleaned #"\s+")]
    (when (and (seq cleaned) (block-bold? block) (>= max-font (* median-font 1.05)))
      (cond
        (is-conditional-hook-heading? text block median-font) true
        (= 1 (count words))
        (and (block-bold? block)
             (>= max-font (* median-font 1.05))
             (Character/isUpperCase ^char (first cleaned))
             (some #(Character/isLowerCase ^char %) (subs cleaned 1)))
        (<= title-case-min-words (count words) title-case-max-words)
        (boolean (re-matches title-case-word-re cleaned))
        :else false))))

(defn is-meta-box-heading?
  [text]
  (contains? meta-box-headings (str/upper-case (strip-glyphs text))))

(defn is-reward-box-heading?
  [text]
  (let [normalized (-> text strip-glyphs (str/split #"\s+") (->> (str/join " ")) str/upper-case)]
    (or (str/starts-with? normalized "RÉCOMPENSE")
        (str/starts-with? normalized "RECOMPENSE"))))

(defn is-narrow-box-heading?
  [text]
  (or (is-meta-box-heading? text) (is-reward-box-heading? text)))

(defn is-credits-heading?
  [text]
  (= (str/upper-case (strip-glyphs text)) "CRÉDITS"))

(defn is-editorial-credits-text?
  [text]
  (let [lowered (str/lower-case (strip-glyphs text))]
    (some #(str/includes? lowered %) editorial-credits-markers)))

(defn is-editorial-credits-block?
  [block]
  (is-editorial-credits-text? (:text block)))

(defn is-all-caps-heading-text?
  [text]
  (let [first-line (-> text strip-glyphs (str/split #"\n") first str/trim)]
    (boolean (re-matches all-caps-re first-line))))

(defn normalize-section-title
  [text]
  (let [stripped (strip-glyphs text)]
    (if (and (str/includes? stripped "\n") (is-all-caps-heading-text? stripped))
      (str/join " " (str/split stripped #"\s+"))
      stripped)))

(defn is-chapter-heading?
  [text]
  (boolean (re-find chapter-re (strip-glyphs text))))

(defn is-list-item-block?
  [block]
  (let [text (strip-glyphs (:text block))]
    (cond
      (not (seq text)) false
      (re-find list-item-marker-re text) true
      (and (block-bold? block) (<= (count (str/split text #"\s+")) 6)) false
      (and (re-matches list-item-name-re text) (<= (count text) 80)) true
      :else false)))

(defn heading-visual-tier
  "Classifie le style : meta, chapter, banner, subordinate, other."
  [text block {:keys [median-font page]}]
  (cond
    (is-meta-box-heading? text) "meta"
    (and page (is-page-banner-heading? text block page)) "banner"
    (is-chapter-heading? text) "chapter"
    (is-title-case-heading? text block median-font) "subordinate"
    :else "other"))

(defn is-drop-cap-false-heading?
  [block page-blocks block-idx]
  (let [text (str/trim (:text block))]
    (and (= 1 (count text))
         (Character/isUpperCase ^char (first text))
         (< (inc block-idx) (count page-blocks))
         (let [nxt (-> (get-in page-blocks [(inc block-idx) :text]) str/trim)]
           (and (seq nxt) (Character/isLowerCase ^char (first nxt)))))))

(defn font-transition-heading?
  "Signal typo corps → titre entre blocs consécutifs."
  [block prev-block page median-font page-blocks block-idx]
  (let [text (str/trim (:text block))
        max-font (block-max-font block)
        is-bold (block-bold? block)
        prev-max (when prev-block (block-max-font prev-block))]
    (and (seq text)
         (<= (count text) 120)
         (<= (count (str/split text #"\s+")) 14)
         (not (is-drop-cap-false-heading? block page-blocks block-idx))
         (not (is-vertical-running-header? block page))
         (not (is-decorative-spread-title? block page median-font))
         (not (and prev-block
                   (is-spread-title-pair? prev-block block page median-font)))
         (or (is-chapter-heading? text)
             (is-meta-box-heading? text)
             (is-reward-box-heading? text)
             (is-title-case-heading? text block median-font)
             (and (re-matches numbered-heading-re text)
                  (or is-bold (>= max-font (* median-font 1.05))))
             (and (re-matches all-caps-re text)
                  (>= (count text) 4)
                  (>= max-font median-font))
             (and is-bold
                  (>= max-font (* median-font 1.15))
                  (<= 3 (count text) 80))
             (and prev-block
                  is-bold
                  (>= max-font (* (or prev-max max-font) 1.1))
                  (<= (count (str/split text #"\s+")) 14)
                  (<= (count text) 120))))))

(defn heading-level
  [text block median-font]
  (let [tier (heading-visual-tier text block {:median-font median-font})]
    (cond
      (#{"meta" "chapter" "banner"} tier) 1
      (= "subordinate" tier) 2
      :else (let [numbered (re-matches numbered-heading-re text)]
              (if numbered
                (min 4 (+ 1 (count (str/split (nth numbered 1) #"\."))))
                (cond
                  (>= (block-max-font block) (* median-font 1.3)) 1
                  (>= (block-max-font block) (* median-font 1.15)) 2
                  :else 3))))))

(defn find-block
  [pages page-number block-index]
  (some (fn [page]
          (when (= (:page-number page) page-number)
            (some #(when (= (:block-index %) block-index) %)
                  (:blocks page))))
        pages))

(defn same-page-caps-parent-id
  [page-num block sections anchors pages]
  (let [page (some #(when (= (:page-number %) page-num) %) pages)
        median (page-median-font (:blocks page))]
  (loop [best-id nil best-y -1.0 idx 0]
    (if (>= idx (count sections))
      best-id
      (let [section (nth sections idx)
            anchor (nth anchors idx)
            parent-block (find-block pages (first anchor) (second anchor))]
        (recur
         (if (and (= (first anchor) page-num)
                  parent-block
                  (not (#{"chapter" "banner" "meta"}
                        (heading-visual-tier (:title section) parent-block
                                             {:median-font median :page page})))
                  (is-all-caps-heading-text? (:title section))
                  (is-in-column-band? block parent-block)
                  (< (get-in parent-block [:bbox :y0]) (get-in block [:bbox :y0]))
                  (<= (- (get-in block [:bbox :y0]) (get-in parent-block [:bbox :y1]))
                      caps-subordinate-max-gap)
                  (> (get-in parent-block [:bbox :y0]) best-y))
           (:id section)
           best-id)
         (if (and parent-block (> (get-in parent-block [:bbox :y0]) best-y))
           (get-in parent-block [:bbox :y0])
           best-y)
         (inc idx)))))))
